src/transform/src/reduce_elision.rs

```diff
@@ -35,48 +36,83 @@ impl crate::Transform for ReduceElision {
-                    .collect_vec();
+    pub fn action(&self, relation: &mut MirRelationExpr, derived: DerivedView) {
+        let mut todo = vec![(relation, derived)];
+        while let Some((expr, view)) = todo.pop() {
+            let mut replaced = false;
+            if let MirRelationExpr::Reduce {
+                input,
+                group_key,
+                aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = expr
+            {
+                let input_type = view
+                    .last_child()
+                    .value::<RelationType>()
+                    .expect("RelationType required")
+                    .as_ref()
+                    .expect("Expression not well-typed");
+                let input_keys = view
+                    .last_child()
+                    .value::<UniqueKeys>()
+                    .expect("UniqueKeys required");
+
+                if input_keys.iter().any(|keys| {
+                    keys.iter()
+                        .all(|k| group_key.contains(&mz_expr::MirScalarExpr::Column(*k)))
+                }) {
+                    let map_scalars = aggregates
+                        .iter()
+                        .map(|a| a.on_unique(input_type))
+                        .collect_vec();
+
+                    let mut result = input.take_dangerous();
 
-                let mut result = input.take_dangerous();
+                    let input_arity = input_type.len();
 
-                // Append the group keys, then any `map_scalars`, then project
-                // to put them all in the right order.
-                let mut new_scalars = group_key.clone();
-                new_scalars.extend(map_scalars);
-                result = result.map(new_scalars).project(
-                    (input_type.column_types.len()
-                        ..(input_type.column_types.len() + (group_key.len() + aggregates.len())))
-                        .collect(),
-                );
+                    // Append the group keys, then any `map_scalars`, then project
+                    // to put them all in the right order.
+                    let mut new_scalars = group_key.clone();
+                    new_scalars.extend(map_scalars);
+                    result = result.map(new_scalars).project(
+                        (input_arity..(input_arity + (group_key.len() + aggregates.len())))
+                            .collect(),
+                    );
 
-                *relation = result;
+                    *expr = result;
+                    replaced = true;
+
+                    // // NB: The following is borked because of smart builders.
+                    // if let MirRelationExpr::Project { input, .. } = expr {
+                    //     if let MirRelationExpr::Map { input, .. } = &mut **input {
+                    //         todo.push((&mut **input, view.last_child()))
+                    //     }
+                    // }
```

Comment:
This is annoying and worth discussing. Without some additional support for post-order visitation, we seemingly need to terminate the pre-order visitation here as we may corrupt the child state in ways we cannot reverse out. Doing a post-order traversal is probably not too hard, just not in love with them. :)